Add fourth Zephyr blog post (#2620)
pulumi/pulumi-hugo: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md

---
title: "IaC Recommended Practices: Using Stack References"
date: 2023-03-31
meta_desc: This post discusses the use of stack references to share information among multiple Pulumi projects, and the recommended practices around their use.
meta_image: meta.png
authors:
    - aaron-kao
    - christian-nunciato
    - scott-lowe
tags:
    - best-practices
    - cloud-engineering
    - zephyr
---

This is the fourth post in a series of blog posts focused on Zephyr Archaeotech Emporium---our fictional company---and their use of Pulumi to manage their online retail store. In the first three posts, you saw how Zephyr's initial use of Pulumi changed as the company grew, and how the use of short-lived per-developer stacks helped Zephyr's application development team meet the demands of a rapidly growing company. This post is a complement to the previous post on structuring Pulumi projects, discussing how Zephyr uses Stack References to link their projects together and sharing some recommended practices around the use of Stack References.<!--more-->

> As you may have read in previous Zephyr posts, the ultimate goal of the Zephyr series is to share recommended practices on the use of Pulumi to manage your infrastructure and application resources (using our fictional company and a fairly complex containerized application as the use case). However, the series exposes those recommended practices over time---not all right away, and not without also discussing the context for the recommendations. This is deliberate, demonstrating how "point-in-time" recommendations change based on the needs of the company and its requirements.

Author: Which portions, in particular, create that impression? (That will help me target the rewording appropriately.)

> Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published---this will get updated as new posts are published):

Reviewer: ```suggestion
Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published and will be updated as new posts are published):
```

Author: Addressed in f28ec13c

* [IaC Recommended Practices: Code Organization and Stacks](/blog/iac-recommended-practices-code-organization-and-stacks/)
* [IaC Recommended Practices: Developer Stacks and Git Branches](/blog/iac-recommended-practices-developer-stacks-git-branches/)
* [IaC Recommended Practices: Structuring Pulumi Projects](/blog/iac-recommended-practices-structuring-pulumi-projects/)
* **IaC Recommended Practices: Using Stack References** (you are here)
* IaC Recommended Practices: Local Testing with Pulumi
* IaC Recommended Practices: Evolving the Application
* IaC Recommended Practices: Adding Pulumi Deployments
* IaC Recommended Practices: Refactoring for Reuse

## Reviewing Zephyr's current status

> As this post is somewhat of a continuation of the previous post, Zephyr's use of Pulumi remains the same as it did at the end of the third post. That is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect or portion of the overall set of requirements for Zephyr's online store:

Reviewer: ```suggestion
This post is a continuation of the previous post in the series: Zephyr's infrastructure footprint and Pulumi codebase remains the same, that is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect of the infrastructure for Zephyr's online store:
```

Author: Incorporated in f23ec13c

> * The "zephyr-infra" project handles base infrastructure (VPC and related constructs).

Author: It's too late to rename the repo now, but it's not to say that we can't show a justification for renaming the project (and the repo) later (this is a supported operation, after all). As for why the 3 repos/projects...well, that was the point of the previous post. :-) I can briefly summarize that again here just for the sake of completeness.

> * The "zephyr-infra" project handles base infrastructure (VPC and related constructs).

Author: Fixed in f28ec13c

> * The "zephyr-infra" project handles base infrastructure (VPC and related constructs).

Reviewer 2: I do think it would be nice to link these projects to their respective repos, though. What do you think? Should we not?

* The "zephyr-app" project handles _only_ the deployment of the online store services onto Kubernetes, layering on top of the platform provided by the "zephyr-k8s" project.

{{% notes %}}
You can continue to use the `multi-project` tag in the GitHub repositories to see the state of Zephyr's code and projects for this blog post.
{{% /notes %}}

It's clear that there are key dependencies across projects. The "zephyr-k8s" project needs to know the VPC ID and the subnet IDs from the base infrastructure project. Similarly, the "zephyr-app" project needs to have access to the Kubernetes cluster details from the "zephyr-k8s" project. As the Zephyr team went about building the code that is now in use in their multi-project architecture, they needed to determine how best to handle these cross-project dependencies. The Zephyr team knew that hard-coding output values from one project as configuration values in another stack was not the ideal way (Pulumi **strongly** recommends against this approach), but what was the best solution to use? The answer to that question is _stack references_.

> ## Linking stacks

Author: Addressed in f28ec13c

> For those that are unfamiliar, stack references are a feature in Pulumi that allows you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [here in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).

Reviewer: ```suggestion
Stack references are a feature in Pulumi that allow you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).
```

Author: Partially addressed in f28ec13c. My English teacher wife and I agree that "allows" is correct, but I'll likely reword this in a future change just to avoid any confusion or disagreement. :-)

Author: Reworded completely in 17aacd6f

> Using stack references, the Zephyr team could make the necessary connections they needed:

Author: Fixed in f28ec13c

> * The code handling the Kubernetes cluster could reference the VPC ID and subnet IDs from the base infrastructure layer.

Reviewer: ```suggestion
* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.
```

Author: Addressed in f28ec13c

> * Similarly, the code that was responsible for deploying the online store services could reference the necessary Kubernetes connection details for the Kubernetes cluster.

Reviewer: ```suggestion
* Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (kubeconfig) for the Kubernetes cluster.
```

Author: Fixed in f28ec13c

> It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Fortunately, adding stack outputs after the fact is easy and generally has no impact whatsoever on existing infrastructure.

Reviewer: We should either omit "generally" or explain the exceptions here. (I believe it never has an effect, but maybe there's something I'm not remembering.)

Author: Done, taken care of in f28ec13c

> Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.

Reviewer: ```suggestion
Additionally, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.
```

Author: Handled by f28ec13c

> Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.

Reviewer 3: we also have this in our style guide https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md

> Additionally, any value retrieved via a stack reference is treated as an `Output`, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.

Author: Fixed in 17aacd6f

> While stack references are generally straightforward, there are some recommended practices to observe around the use of stack outputs and stack references:

Reviewer: ```suggestion
While stack references are conceptually straightforward and easy to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:
```

1. With regard to stack outputs, **only export what is needed.** If there's a need for it to be accessed from outside the stack, export it; otherwise, don't. It's easy to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're only modifying the stack object itself).

> 2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.

Author: @ringods @pierskarsenbarg Since you folks brought up the concern regarding minimize stack references and calls to read values, can you confirm if the concern is in stack outputs or stack references? Based on the comments in Slack, I believe it is the reading stack (and therefore stack references are the concern).

> 2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.

Author: Resolved via Slack discussions (see threads in `#project-zephyr` and `#questions`).

> 3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of additional work required in this situation.

Author: @ringods @pierskarsenbarg Pinging you two again here. I read the exchange in Slack as needing to build an interface in the reading stack, but I could certainly see the need to build an interface/struct in the source stack as well. Can you confirm based on your understanding?

> 3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in only a single call to the backend. Be aware, however, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of additional work required in this situation.

Author: After some lengthy Slack discussions, I think this has been resolved.

> 4. In line with parameterizing as much of your code as possible, **also be sure to parameterize your stack references.** A stack reference is built using an organization name, a project name, and a stack name. Don't hardcode these values; instead, use configuration values to allow the users to specify from which source stack(s) the values will be referenced.

Reviewer: I can't think of a use case where organization would need to be templates, but stack name definitely should be. It might be nice to show a default value that is the current stack name, e.g.

```
const stack = new pulumi.Config().getStack().split("/")[2];
const stackRef = new pulumi.StackReference(`my-organization/workshop-aws-proton-base-infra/${stack}`);
```

Double check the above code in an actual program before publishing - I can't vouch for its correctness, but it's something like that.

Author: To me, it seems entirely plausible that a company could have two organizations, with users that are members of both organizations, and there be a need to read information across the organizational boundary via a stack reference (two different business units, perhaps?).

Reviewer 2: +1 to add some code here to show what "parameterize your stack references" looks like.

Author: Code added in 17aacd6f, along with references to the code in other parts of the post

> ## Examining Zephyr's use of stack references

Author: I'd really like readers to be able to see the highlighted code in the context of the broader Pulumi program. How strongly do you feel about putting code inline into the blog post?

> ## Examining Zephyr's use of stack references

Author: Code samples added in 17aacd6f

## Linking stacks with stack references

Stack references are a feature in Pulumi that allows you to access output values from one stack in another stack. In other words, stack references offer a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).

Using stack references, the Zephyr team can make the necessary connections they needed:

* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.

> * Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (the Kubeconfig) for the Kubernetes cluster.

Author: Code was added toward the end in 17aacd6f and I've put a reference to that code here

It's worthwhile to note that any information that needs to be accessible from another stack via a stack reference must be exported as a stack output in the source stack. If you don't mark it as a stack output, then it can't be used in a stack reference. Adding stack outputs after the fact requires little effort and has no impact on existing infrastructure.

> Also, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some extra work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps considerably in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.

Author: Done, fixed in 17aacd6f

While stack references are conceptually straightforward and not difficult to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:

> 1. When it comes to stack outputs, **export what is needed.** If it needs to be accessed from outside the stack, export it; otherwise, don't. It's simple to add more stack outputs after the fact, and as has been mentioned already this is done with a quick `pulumi up` that has no affect on existing infrastructure (you're modifying the stack object itself).

Reviewer 2: >If it needs to be accessed from outside the stack, export it; otherwise, don't.

It occurred to me that after reading this, I might wonder, Why not? Is there any harm in doing so? There probably isn't, unless what you're exporting is somehow potentially sensitive of course. The "why" here might therefore be something along the lines of the same "why" that encourages encapsulation in a public API: expose what the API user cares about or is likely to need, hide all the rest as implementation detail (which ultimately leaves you with more ability to change the underlying implementation without affecting the API).

Author: I've added some text, borrowing from your suggestion, in 17aacd6f

2. **Be judicious in the use of stack references.** Each stack reference is a call to the backend to read a value, which adds some level of latency. Using lots of stack references (think more than 20 or so) can introduce notable latency in Pulumi operations.

> 3. **If you need lots of stack references, use a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.

Reviewer 2: I might tweak this slightly to clarify we're talking about the exporting, not the consuming (but up to you -- feel free to ignore if you don't think it adds anything):

```suggestion
3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
```

It might also be worth mentioning that this approach can also make it a little easier to codify the API contract between stacks and their consumers (vs. that contract consisting of "a bunch of optional strings"). 

Relatedly, we might want to mention that outputs are expected to be primitive/serializable types -- string, bool, object, etc. Occasionally folks think they can export whole resource instances (as you would in a typical program) and then realize this doesn't quite work, and are confused. (The docs may also mention this, though.)

Author: In 17aacd6f I've reworded per your suggestion, but there is still an outstanding question as to whether it is the stack outputs that are "costly" or if it is the stack references reading those outputs that's "costly" (from a latency perspective). I've asked @ringods & @pierskarsenbarg to weigh in on that, as I'm not sure.

Author: I think this has been resolved (refer to Slack discussions in #project-zephyr and #questions).

* In the Kubernetes platform stack, the Zephyr team [parameterized the values](https://github.com/pulumi/zephyr-k8s/blob/multi-project/index.ts#L10-L12) needed for the stack reference. This particular approach, by the way, is key to ensuring [the per-developer stacks](/blog/iac-recommended-practices-developer-stacks-git-branches/) to which the Zephyr team has grown accustomed are accommodated (each developer needs to specify the correct organization, project, and stack name).
* As with the base infrastructure stack, [only the key value needed by other stacks](https://github.com/pulumi/zephyr-k8s/blob/multi-project/index.ts#L36-L37) is exported.
* Finally, in the application stack, the [stack reference values are again parameterized](https://github.com/pulumi/zephyr-app/blob/multi-project/infra/index.ts#L7-L9), and the Kubeconfig---necessary for the application stack to deploy onto the provisioned Kubernetes cluster---is [referenced via a stack reference](https://github.com/pulumi/zephyr-app/blob/multi-project/infra/index.ts#L11-L13).

{{% notes %}}

> All of the GitHub links in the paragraphs above reference the `multi-project` tag in each repository, which has the code as of this blog post (and the earlier blog post).

Author: > I link to link these references
To which references are you referring? The blog post references? Or the GitHub references? (I think the latter, if I'm reading your comment correctly.)

> All of the GitHub links in the paragraphs above reference the `multi-project` tag in each repository, which has the code as of this blog post (and the earlier blog post).

Reviewer 2: You know what, I'm sorry -- I must've misread this. I think I read it as "The code above refers to the `multi-project` tag in the repository..." so my suggestion was just to link "multi-project tag" to the tag in the repo. :ignoreme:
